### PR TITLE
GSoC 2021 : Unit Tests for User Controller added.

### PR DESCRIPTION
**Pull request by adityaofficial10**
The following additions have been made:
-  Unit tests for all the user controllers have been added.
- The structure for the controllers has been changed a bit.

Structure Change: 
The async-await structure has been used to replace then calls as I think that should be the way to move forward.

@katjaq  @r03ert0 
Kindly have a look.
I don't think something is breaking but kindly verify once..

### CODE REVIEW

File: test/unit/mriController.test.js [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+            status: sinon.stub().returns({ send: sinon.stub().returns({ end: resSpy })}),
+        };
+        await mriController.validatorPost(req, res, () => {});
+        assert.strictEqual(reqSpy.callCount, 1);
+        assert.strictEqual(urlSpy.callCount, 1);
+        assert.strictEqual(resSpy.callCount, 0);
+        sinon.restore();
+    });
+
+    it('download MRI function should download the file when correct input is given', async function() {
```
Comment (r03ert0) [inline test/unit/mriController.test.js:68]: A call to the mri() function does not download the file, but the entry in the mongoDB associated with that file. Change the test name to "should get the MRI record associated with the URL"

File: test/unit/mriController.test.js [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+        let res = {
+            render: sinon.spy(),
+            status: sinon.stub().returns({ send: sinon.stub().returns({ end: resSpy })}),
+        };
+        await mriController.reset(req, res);
+        assert.notDeepStrictEqual(resSpy.calledOnce, true);
+        sinon.restore();
+    });
+
+    it('appMriGet function() should fetch the MRI as directed when the URL is correct', async function() {
```
Comment (r03ert0) [inline test/unit/mriController.test.js:120]: this call may leave a hanging entry in the downloadQueue, which could fail later on. Same problem as in PR #287.

File: test/unit/mriController.test.js [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+        let req = {
+            db: db,
+            query: {
+                url: 'https://brainbox.pasteur.fr/mri?url=https://s3.amazonaws.com/fcp-indi/data/Projects/ABIDE_Initiative/Outputs/freesurfer/5.1/CMU_a_0050642/mri/T1.mgz',
```
Comment (r03ert0) [inline test/unit/mriController.test.js:124]: The url is incorrect. It should be https://s3.amazon.com/... etc.
(https://brainbox.pasteur.fr/mri?url=https://s3.amazonws.com/... etc. is the URL for the webpage associated to the MRI file sitting at AWS S3)

File: test/unit/mriController.test.js [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+        assert.notDeepStrictEqual(jsonSpy.calledOnce, true);
+        sinon.restore();
+    });
+
+    it('apiMriPost function() should work correctly and make the right calls when input is correct', async function() {
+        let req = {
+            db: db,
+            body: {},
+            query: {
+                url: 'https://brainbox.pasteur.fr/mri?url=https://s3.amazonaws.com/fcp-indi/data/Projects/ABIDE_Initiative/Outputs/freesurfer/5.1/CMU_a_0050642/mri/T1.mgz',
```
Comment (r03ert0) [inline test/unit/mriController.test.js:185]: incorrect URL. Drop the brainbox.pasteur.fr part

File: test/unit/mriController.test.js [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+        assert.notDeepStrictEqual(jsonSpy.calledOnce, true);
+        sinon.restore();
+    });
+
+    it('apiMriPost function() should throw an error when input is incorrect', async function() {
+        let req = {
+            db: db,
+            body: {},
+            query: {
+                url: '',
```
Comment (r03ert0) [inline test/unit/mriController.test.js:214]: a POST to url: '' will return a paginated list of available MRIs (in this case, page=1), and not an error. The test should be split in two: One for testing that paginated MRI lists are working, and one for testing that an incorrect URL does fail. Use an invalid URL, like `url: "invalid_url"`.

File: test/unit/mriController.test.js [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+        assert.strictEqual(resSpy.callCount, 0);
+        sinon.restore();
+    });
+
+    it('token function() should return the token with correct credentials', async function() {
+        let req = {
+            db: db,
+            body: {},
+            query: {
+                url: 'https://brainbox.pasteur.fr/mri?url=https://s3.amazonaws.com/fcp-indi/data/Projects/ABIDE_Initiative/Outputs/freesurfer/5.1/CMU_a_0050642/mri/T1.mgz',
```
Comment (r03ert0) [inline test/unit/mriController.test.js:294]: incorrect URL.

File: test/unit/mriController.test.js [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+        assert.strictEqual(jsonSpy.callCount, 1);
+        sinon.restore();
+    });
+
+    it('token function() should redirect with incorrect credentials', async function() {
+        let req = {
+            db: db,
+            body: {},
+            query: {
+                url: 'https://brainbox.pasteur.fr/mri?url=https://s3.amazonaws.com/fcp-indi/data/Projects/ABIDE_Initiative/Outputs/freesurfer/5.1/CMU_a_0050642/mri/T1.mgz',
```
Comment (r03ert0) [inline test/unit/mriController.test.js:322]: incorrect URL

File: test/unit/user.controller.test.js [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+            assert.strictEqual(res.send.callCount, 0);
+            sinon.restore();
+        });
+    });
+
+    describe('user function() ', function () {
+        it('should return the requested user with valid input', async function () {
+            let req = {
+                user: {
+                    username: 'anyone'
```
Comment (r03ert0) [inline test/unit/user.controller.test.js:27]: Brainbox uses "nickname" for "username". Please check the "user" collection in the Mongo DB

File: test/unit/user.controller.test.js [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+                },
+                params: {
+                    userName: 'foo'
```
Comment (r03ert0) [inline test/unit/user.controller.test.js:30]: should be "nickname"

File: test/unit/user.controller.test.js [hunk 10]
```diff
@@ -0,0 +1,394 @@
...
+            };
+            await userController.user(req, res);
+            assert.strictEqual(res.render.callCount, 1);
+            sinon.restore();
+        });
+
+        it('should throw no user message when the input is invalid', async function() {
+            let req = {
+                user: {
+                    username: 'anyone'
```
Comment (r03ert0) [inline test/unit/user.controller.test.js:53]: should be "nickname"

File: test/unit/user.controller.test.js (hunk 8 above)
Comment (adityaofficial10) [reply test/unit/user.controller.test.js:27]: Resolved..

**Comment by r03ert0** [comment]
thank you very much @adityaofficial10!

**Pull request merged**